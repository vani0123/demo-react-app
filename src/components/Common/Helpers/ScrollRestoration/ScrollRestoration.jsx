import { useEffect } from 'react';
import { useLocation } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { PropTypes } from 'prop-types';

import { CommonUtil } from '../../../../utils';
import { commonAction } from '../../../../redux/actions';

/**
 * ScrollRestoration component, responsible for restoring page scroll on page navigation.
 *
 * It also allow on demand scroll restoration on calling restoreScroll action defined in
 * common redux actions.
 *
 * @example
 * let isRestoring = false;
 *
 * const actions = {
 *   restoreScrollSuccess: () => {
 *    isRestoring = false
 *   }
 * };
 *
 * <ScrollRestoration isRestoring={isRestoring} actions={actions}>
 *   <App />
 * <ScrollRestoration/>
 */
const ScrollRestoration = ({ children, isRestoring, actions }) => {
  const { pathname } = useLocation();

  useEffect(() => {
    /**
     * Calls scrollToTop method defined in common utils when even page url and isRestoring flag is changed.
     * This method is responsible for restoring page scroll.
     */
    CommonUtil.scrollToTop();
    actions.restoreScrollSuccess();
  }, [pathname, isRestoring]);

  return children;
};

/**
 * mapStateToProps method maps isRestoring redux state properties to props.
 *
 * This will ensure any change to this properties will trigger re-render of this component.
 */
function mapStateToProps(state) {
  return {
    isRestoring: state.common.scrollIndicator.isRestoring
  };
}

/**
 * mapDispatchToProps method binds common redux actions to component's props.
 *
 * This will enable ScrollRestoration component to make calls to common actions directly
 * from props.
 *
 * eg -
 * this.props.actions.restoreScrollSuccess(params)
 */
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({ ...commonAction }, dispatch)
  };
}

/**
 * Default PropTypes definition for ScrollRestoration component.
 *
 * children - Defines the initial value for child component that is wrapped by ScrollRestoration component
 */
ScrollRestoration.defaultProps = {
  children: null
};

/**
 * PropTypes definition for ScrollRestoration component.
 *
 * children - Defines the child component that will wrapped by ScrollRestoration component.
 * actions.restoreScrollSuccess - Redux action to update the scroll restoration state.
 * isRestoring - Defined the page restoration state, sets to true when page scroll restoration is in progress.
 */
ScrollRestoration.propTypes = {
  children: PropTypes.shape({}),
  isRestoring: PropTypes.bool.isRequired,
  actions: PropTypes.shape({
    restoreScrollSuccess: PropTypes.func.isRequired
  }).isRequired
};

/**
 * This component is exported with wrapper of redux connect
 * HOC and make the use of actions and features provided by them.
 */
export default connect(mapStateToProps, mapDispatchToProps)(ScrollRestoration);
