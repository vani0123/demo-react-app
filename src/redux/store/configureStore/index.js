import { createStore, applyMiddleware, compose } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import createSagaMiddleware from 'redux-saga';
import rootReducer from '../../reducers';
import { initSagas } from '../../sagas/initSagas';

export default function configureStore(initialState) {
  const shouldHideDevTool = process.env.NODE_ENV === 'production'
    || typeof document === 'undefined'
    || !window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

  const composeEnhancers = shouldHideDevTool
    ? compose
    : window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__; // add support for Redux dev tools

  const sagaMiddleware = createSagaMiddleware({
    onError(error) {
      setImmediate(() => {
        throw error;
      });
    }
  });
  const store = createStore(
    rootReducer,
    initialState,
    composeEnhancers(
      applyMiddleware(sagaMiddleware, reduxImmutableStateInvariant())
    )
  );

  initSagas(sagaMiddleware);

  return store;
}
