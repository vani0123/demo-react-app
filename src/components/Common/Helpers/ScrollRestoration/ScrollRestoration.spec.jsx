import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from '../../../../redux/store/configureStore';
import ScrollRestoration from './ScrollRestoration';

jest.mock('react-router-dom', () => ({
  useLocation: () => ({
    pathname: 'mockedPath'
  })
}));

const setup = () => {
  const store = configureStore();
  return mount(
    <Provider store={store}>
      <ScrollRestoration />
    </Provider>
  );
};

describe('The ScrollRestoration component', () => {
  it('should renders without crashing', () => {
    const wrapper = setup();
    expect(wrapper.find('ScrollRestoration').exists()).toBe(true);
  });

  it('should have valid snapshot', () => {
    const wrapper = setup(true);
    expect(wrapper).toMatchSnapshot();
  });
});
