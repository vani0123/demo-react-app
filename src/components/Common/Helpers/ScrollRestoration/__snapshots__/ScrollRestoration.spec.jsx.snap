// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ScrollRestoration component should have valid snapshot 1`] = `ReactWrapper {}`;
